File: bot-api/dotnet/src/mapper/BulletStateMapper.cs
using System.Collections.Generic;

namespace Robocode.TankRoyale.BotApi.Mapper
{
  public sealed class BulletStateMapper
  {
    public static BulletState Map(Schema.BulletState source)
    {
      return new BulletState(
        source.BulletId,
        source.OwnerId,
        source.Power,
        source.X,
        source.Y,
        source.Direction,
        source.Speed,
        Color.FromRgbInt(source.Color)
      );
    }

    public static ISet<BulletState> Map(IEnumerable<Schema.BulletState> source)
    {
      var bulletStates = new HashSet<BulletState>();
      foreach (var bulletState in source)
      {
        bulletStates.Add(Map(bulletState));
      }
      return bulletStates;
    }
  }
}

File: bot-api/typescript/src/mapper/BulletStateMapper.ts
import { BulletState as SchemaBulletState } from "../protocol/schema.js";
import { BulletState } from "../BulletState.js";
import { ColorUtil } from "../util/ColorUtil.js";

/** Maps schema BulletState to API BulletState. */
export class BulletStateMapper {
  static map(s: SchemaBulletState): BulletState {
    return new BulletState(s.bulletId, s.ownerId, s.power, s.x, s.y, s.direction, ColorUtil.fromHexColor(s.color ?? null));
  }

  static mapCollection(states: SchemaBulletState[]): BulletState[] {
    return states.map(BulletStateMapper.map);
  }
}
